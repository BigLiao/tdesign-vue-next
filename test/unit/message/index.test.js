import { mount } from '@vue/test-utils';
import { Message } from '@/src/message/index.ts';
import Button from '@/src/button/index.ts';
import TIconClose from '@/src/icon/close';
import TIconMore from '@/src/icon/more';
import TIconInfoCircleFilled from '@/src/icon/info-circle-filled';
import TIconCheckCircleFilled from '@/src/icon/check-circle-filled';
import TIconErrorCircleFilled from '@/src/icon/error-circle-filled';
import TIconHelpFill from '@/src/icon/help-circle-filled';
import TIconLoadingFill from '@/src/icon/loading';

// every component needs four parts: props/events/slots/functions.
describe('Message', () => {
  // test props api
  describe(':props', () => {
    const THEME_MAP = {
      info: TIconInfoCircleFilled,
      success: TIconCheckCircleFilled,
      warning: TIconErrorCircleFilled,
      error: TIconErrorCircleFilled,
      question: TIconHelpFill,
      loading: TIconLoadingFill,
    };
    const THEME_LIST = Object.keys(THEME_MAP);

    it('pure message contains two classes', () => {
      const wrapper = mount(Message);
      const classes = wrapper.classes();
      expect(classes).toContain('t-message');
      expect(classes).toContain('t-is-info');
      expect(wrapper.contains(TIconClose)).toBe(false);
      expect(wrapper).toMatchSnapshot();
    });

    it(`:theme ${THEME_LIST.join()}`, () => {
      THEME_LIST.forEach((t) => {
        const wrapper = mount({
          render() {
            return <Message theme={t}></Message>;
          },
        });
        const msg = wrapper.find(Message);
        expect(msg.classes()).toContain(`t-is-${t}`);
        expect(msg.contains(THEME_MAP[t]));
        expect(wrapper).toMatchSnapshot();
      });
    });

    it(':closeBtn is true, render default close button.', () => {
      const wrapper = mount({
        render() {
          return <Message closeBtn={true}></Message>;
        },
      });
      const msg = wrapper.find(Message);
      expect(msg.contains('.t-message-close')).toBe(true);
      expect(msg.contains(TIconClose)).toBe(true);
      expect(wrapper).toMatchSnapshot();
    });

    it(':closeBtn is a tring, equal "关闭".', () => {
      const wrapper = mount({
        render() {
          return <Message closeBtn='关闭'></Message>;
        },
      });
      const msg = wrapper.find(Message);
      const close = msg.find('.t-message-close');
      expect(close.isEmpty()).toBe(false);
      expect(close.text()).toBe('关闭');
      expect(wrapper).toMatchSnapshot();
    });

    it(':closeBtn is a function, () => VNode.', () => {
      const wrapper = mount({
        render() {
          return (
            <Message
              closeBtn={() => <b class='t-message-close-content'>x</b>}
            ></Message>
          );
        },
      });
      const msg = wrapper.find(Message);
      const close = msg.find('.t-message-close-content');
      expect(close.isEmpty()).toBe(false);
      expect(close.html()).toBe('<b class="t-message-close-content">x</b>');
      expect(wrapper).toMatchSnapshot();
    });

    it(':icon is false', () => {
      const wrapper = mount({
        render() {
          return <Message icon={false} ></Message>;
        },
      });
      const msg = wrapper.find(Message);
      expect(msg.contains('.t-icon')).toBe(false);
      expect(wrapper).toMatchSnapshot();
    });

    it(':icon is a funtion, () => TIconMore', () => {
      const wrapper = mount({
        render() {
          return (
            <Message
              icon={() => <TIconMore></TIconMore>}
            ></Message>
          );
        },
      });
      expect(wrapper.contains(TIconMore)).toBe(true);
      expect(wrapper).toMatchSnapshot();
    });

    it(':default is a funtion, () => <b>这是重要信息</b>', () => {
      const wrapper = mount({
        render() {
          return (
            <Message
              content={() => <b>这是重要信息</b>}
            ></Message>
          );
        },
      });
      expect(wrapper).toMatchSnapshot();
    });

    it(':default is a string, 这是普通信息', () => {
      const wrapper = mount({
        render() {
          return (
            <Message content='这是普通信息'></Message>
          );
        },
      });
      expect(wrapper).toMatchSnapshot();
    });
  });

  // test events
  describe('@event', () => {
    it('@click-close-btn', async () => {
      const wrapper = mount({
        render() {
          return <Message closeBtn={true}></Message>;
        },
      });
      const msg = wrapper.find(Message);
      await msg.find('.t-icon-close').trigger('click');
      expect(msg.emitted()['click-close-btn']).toBeTruthy();
    });

    it('@duration-end', (done) => {
      const wrapper = mount({
        render() {
          return <Message duration={3000}></Message>;
        },
      });
      const msg = wrapper.find(Message);
      expect(msg.emitted()['duration-end']).toBeFalsy();
      const timer = setTimeout(() => {
        const msg = wrapper.find(Message);
        expect(msg.emitted()['duration-end']).toBeTruthy();
        done();
        clearTimeout(timer);
      }, 3100);
    });
  });

  // test slots
  describe('<slot>', () => {
    it('<default>', () => {
      const wrapper = mount({
        render() {
          return (
            <Message
              {...{
                scopedSlots: {
                  default: () => <Button class='custom-message'>自定义细心内容</Button>,
                },
              }}
            ></Message>
          );
        },
      });
      expect(wrapper.contains(Button)).toBe(true);
      expect(wrapper.contains('.custom-message')).toBe(true);
    });

    it('<icon>', () => {
      const wrapper = mount({
        render() {
          return (
            <Message
              {...{
                scopedSlots: {
                  icon: () => <TIconMore></TIconMore>,
                },
              }}
            ></Message>
          );
        },
      });
      expect(wrapper.contains(TIconMore)).toBe(true);
      expect(wrapper).toMatchSnapshot();
    });

    it('<closeBtn>', () => {
      const wrapper = mount({
        render() {
          return (
            <Message
              {...{
                scopedSlots: {
                  closeBtn: () => <div class='custome-close-btn'>x</div>,
                },
              }}
            ></Message>
          );
        },
      });
      expect(wrapper.contains('.custome-close-btn')).toBe(true);
    });
  });
});
