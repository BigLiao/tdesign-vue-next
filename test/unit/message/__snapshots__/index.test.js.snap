// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Message :props :closeBtn is a function, () => VNode. 1`] = `
<div
  class="t-message t-is-info t-is-closable"
>
  <svg
    class="t-icon t-icon-info-circle-filled"
    version="1.1"
    viewBox="0 0 1024 1024"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M512 96C741.7504576 96 928 282.24954432 928 512 928 741.7504576 741.7504576 928 512 928 282.24954432 928 96 741.7504576 96 512 96 282.24954432 282.24954432 96 512 96ZM544 416L416 416 416 480 480 480 480 736 608 736 608 672 544 672 544 416ZM544 288L480 288 480 352 544 352 544 288Z"
      fill-opacity=".9"
      fill-rule="evenodd"
    />
  </svg>
  <!---->
  <span
    class="t-message-close"
  >
    <b
      class="t-message-close-content"
    >
      x
    </b>
  </span>
</div>
`;

exports[`Message :props :closeBtn is a tring, equal "关闭". 1`] = `
<div
  class="t-message t-is-info t-is-closable"
>
  <svg
    class="t-icon t-icon-info-circle-filled"
    version="1.1"
    viewBox="0 0 1024 1024"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M512 96C741.7504576 96 928 282.24954432 928 512 928 741.7504576 741.7504576 928 512 928 282.24954432 928 96 741.7504576 96 512 96 282.24954432 282.24954432 96 512 96ZM544 416L416 416 416 480 480 480 480 736 608 736 608 672 544 672 544 416ZM544 288L480 288 480 352 544 352 544 288Z"
      fill-opacity=".9"
      fill-rule="evenodd"
    />
  </svg>
  <!---->
  <span
    class="t-message-close"
  >
    关闭
  </span>
</div>
`;

exports[`Message :props :closeBtn is true, render default close button. 1`] = `
<div
  class="t-message t-is-info t-is-closable"
>
  <svg
    class="t-icon t-icon-info-circle-filled"
    version="1.1"
    viewBox="0 0 1024 1024"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M512 96C741.7504576 96 928 282.24954432 928 512 928 741.7504576 741.7504576 928 512 928 282.24954432 928 96 741.7504576 96 512 96 282.24954432 282.24954432 96 512 96ZM544 416L416 416 416 480 480 480 480 736 608 736 608 672 544 672 544 416ZM544 288L480 288 480 352 544 352 544 288Z"
      fill-opacity=".9"
      fill-rule="evenodd"
    />
  </svg>
  <!---->
  <span
    class="t-message-close"
  >
    <svg
      class="t-icon t-icon-close"
      version="1.1"
      viewBox="0 0 1024 1024"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M760.9780096 217.6L806.2328448 262.85483392 557.312 511.744 806.2328448 760.6580096 760.9780096 805.9128448 512.064 556.992 263.17483392 805.9128448 217.92 760.6580096 466.816 511.744 217.92 262.85483392 263.17483392 217.6 512.064 466.496 760.9780096 217.6Z"
        fill-opacity=".9"
        fill-rule="evenodd"
      />
    </svg>
  </span>
</div>
`;

exports[`Message :props :default is a funtion, () => <b>这是重要信息</b> 1`] = `
<div
  class="t-message t-is-info"
>
  <svg
    class="t-icon t-icon-info-circle-filled"
    version="1.1"
    viewBox="0 0 1024 1024"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M512 96C741.7504576 96 928 282.24954432 928 512 928 741.7504576 741.7504576 928 512 928 282.24954432 928 96 741.7504576 96 512 96 282.24954432 282.24954432 96 512 96ZM544 416L416 416 416 480 480 480 480 736 608 736 608 672 544 672 544 416ZM544 288L480 288 480 352 544 352 544 288Z"
      fill-opacity=".9"
      fill-rule="evenodd"
    />
  </svg>
  <b>
    这是重要信息
  </b>
  <span
    class="t-message-close"
  >
    <!---->
  </span>
</div>
`;

exports[`Message :props :default is a string, 这是普通信息 1`] = `
<div
  class="t-message t-is-info"
>
  <svg
    class="t-icon t-icon-info-circle-filled"
    version="1.1"
    viewBox="0 0 1024 1024"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M512 96C741.7504576 96 928 282.24954432 928 512 928 741.7504576 741.7504576 928 512 928 282.24954432 928 96 741.7504576 96 512 96 282.24954432 282.24954432 96 512 96ZM544 416L416 416 416 480 480 480 480 736 608 736 608 672 544 672 544 416ZM544 288L480 288 480 352 544 352 544 288Z"
      fill-opacity=".9"
      fill-rule="evenodd"
    />
  </svg>
  这是普通信息
  <span
    class="t-message-close"
  >
    <!---->
  </span>
</div>
`;

exports[`Message :props :icon is a funtion, () => TIconMore 1`] = `
<div
  class="t-message t-is-info"
>
  <svg
    class="t-icon t-icon-more"
    version="1.1"
    viewBox="0 0 1024 1024"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M576 192.00000000000003C576 227.34622400000003 547.346224 256.00000000000006 512 256.00000000000006 476.653776 256.00000000000006 448 227.34622400000003 448 192.00000000000003 448 156.65377600000002 476.653776 128.00000000000003 512 128.00000000000003 547.346224 128.00000000000003 576 156.65377600000002 576 192.00000000000003ZM576 512C576 547.346224 547.3462240000001 576 512 576 476.653776 576 448 547.346224 448 512 448 476.65377600000005 476.653776 448.00000000000006 512 448.00000000000006 547.346224 448.00000000000006 576 476.653776 576 512ZM576 832C576 867.3462208 547.3462240000001 896 512 896 476.653776 896 448 867.3462208 448 832 448 796.6537792 476.653776 768 512 768 547.3462240000001 768 576 796.6537792 576 832Z"
      fill-rule="evenodd"
    />
  </svg>
  <!---->
  <span
    class="t-message-close"
  >
    <!---->
  </span>
</div>
`;

exports[`Message :props :icon is false 1`] = `
<div
  class="t-message t-is-info"
>
  <!---->
  <!---->
  <span
    class="t-message-close"
  >
    <!---->
  </span>
</div>
`;

exports[`Message :props :theme info,success,warning,error,question,loading 1`] = `
<div
  class="t-message t-is-info"
>
  <svg
    class="t-icon t-icon-info-circle-filled"
    version="1.1"
    viewBox="0 0 1024 1024"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M512 96C741.7504576 96 928 282.24954432 928 512 928 741.7504576 741.7504576 928 512 928 282.24954432 928 96 741.7504576 96 512 96 282.24954432 282.24954432 96 512 96ZM544 416L416 416 416 480 480 480 480 736 608 736 608 672 544 672 544 416ZM544 288L480 288 480 352 544 352 544 288Z"
      fill-opacity=".9"
      fill-rule="evenodd"
    />
  </svg>
  <!---->
  <span
    class="t-message-close"
  >
    <!---->
  </span>
</div>
`;

exports[`Message :props :theme info,success,warning,error,question,loading 2`] = `
<div
  class="t-message t-is-success"
>
  <svg
    class="t-icon t-icon-check-circle-filled"
    version="1.1"
    viewBox="0 0 1024 1024"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M512 96C741.7504576 96 928 282.24954432 928 512 928 741.7504576 741.7504576 928 512 928 282.24954432 928 96 741.7504576 96 512 96 282.24954432 282.24954432 96 512 96ZM690.8655168 353.58818368L464.56401024 579.82818368 328.82684416 444.09785152 283.57201024 489.35268544 464.59134592 670.3720192 509.84617984 625.1171872 736.120352 398.8430176 690.8655168 353.58818368Z"
      fill-opacity=".9"
      fill-rule="evenodd"
    />
  </svg>
  <!---->
  <span
    class="t-message-close"
  >
    <!---->
  </span>
</div>
`;

exports[`Message :props :theme info,success,warning,error,question,loading 3`] = `
<div
  class="t-message t-is-warning"
>
  <svg
    class="t-icon t-icon-error-circle-filled"
    version="1.1"
    viewBox="0 0 1024 1024"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M512 96C741.7504576 96 928 282.24954432 928 512 928 741.7504576 741.7504576 928 512 928 282.24954432 928 96 741.7504576 96 512 96 282.24954432 282.24954432 96 512 96ZM544 672L480 672 480 736 544 736 544 672ZM544 288L480 288 480 608 544 608 544 288Z"
      fill-opacity=".9"
      fill-rule="evenodd"
    />
  </svg>
  <!---->
  <span
    class="t-message-close"
  >
    <!---->
  </span>
</div>
`;

exports[`Message :props :theme info,success,warning,error,question,loading 4`] = `
<div
  class="t-message t-is-error"
>
  <svg
    class="t-icon t-icon-error-circle-filled"
    version="1.1"
    viewBox="0 0 1024 1024"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M512 96C741.7504576 96 928 282.24954432 928 512 928 741.7504576 741.7504576 928 512 928 282.24954432 928 96 741.7504576 96 512 96 282.24954432 282.24954432 96 512 96ZM544 672L480 672 480 736 544 736 544 672ZM544 288L480 288 480 608 544 608 544 288Z"
      fill-opacity=".9"
      fill-rule="evenodd"
    />
  </svg>
  <!---->
  <span
    class="t-message-close"
  >
    <!---->
  </span>
</div>
`;

exports[`Message :props :theme info,success,warning,error,question,loading 5`] = `
<div
  class="t-message t-is-question"
>
  <svg
    class="t-icon t-icon-help-circle-filled"
    version="1.1"
    viewBox="0 0 1024 1024"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M512 96C741.7504576 96 928 282.24954432 928 512 928 741.7504576 741.7504576 928 512 928 282.24954432 928 96 741.7504576 96 512 96 282.24954432 282.24954432 96 512 96ZM544 736L480 736 480 800 544 800 544 736ZM512.2952448 256.64610496C441.57726976 255.489344 377.22819968 301.1206464 357.66394688 370.52669568 354.01781184 383.46170624 352.35546368 401.42816192 352.35546368 415.98780736L352.35546368 415.98780736 416.35546368 415.98780736 416.56729536 408.5808992C416.9809312 400.82021056 417.96989888 392.47948096 419.26346432 387.8904352 430.77765376 347.04275136 468.98897024 319.94628928 511.248512 320.63754496 564.72860672 321.51233984 607.554592 358.4738848 608.34222656 416.42270848 608.88517824 456.36943488 596.93322496 476.61043904 556.124528 517.76842944L556.124528 517.76842944 543.2981664 530.90088384C497.88380224 578.59564288 480 612.83118528 480 672.6246144L480 672.6246144 544 672.6246144 544.23439808 663.0609024C546.13973632 626.16948096 560.01985152 604.73470528 598.90074624 565.52317824L598.90074624 565.52317824 611.51250368 552.613136C655.9401472 505.9730656 673.1166976 472.96825088 672.3363136 415.55290688 671.047328 320.71799424 598.442304 258.05524608 512.2952448 256.64610496Z"
      fill-opacity=".9"
      fill-rule="evenodd"
    />
  </svg>
  <!---->
  <span
    class="t-message-close"
  >
    <!---->
  </span>
</div>
`;

exports[`Message :props :theme info,success,warning,error,question,loading 6`] = `
<div
  class="t-message t-is-loading"
>
  <svg
    class="t-icon t-icon-loading"
    version="1.1"
    viewBox="0 0 1024 1024"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M96 512C96 282.264 282.264 96 512 96 741.736 96 928 282.264 928 512L824 512 824 512C824 339.68715776 684.3128448 200 512 200 339.68715776 200 200 339.68715776 200 512 200 684.3128448 339.68715776 824 512 824L512 928C282.264 928 96 741.736 96 512Z"
      fill-opacity=".9"
      fill-rule="evenodd"
    />
  </svg>
  <!---->
  <span
    class="t-message-close"
  >
    <!---->
  </span>
</div>
`;

exports[`Message :props pure message contains two classes 1`] = `
<div
  class="t-message t-is-info"
>
  <svg
    class="t-icon t-icon-info-circle-filled"
    version="1.1"
    viewBox="0 0 1024 1024"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M512 96C741.7504576 96 928 282.24954432 928 512 928 741.7504576 741.7504576 928 512 928 282.24954432 928 96 741.7504576 96 512 96 282.24954432 282.24954432 96 512 96ZM544 416L416 416 416 480 480 480 480 736 608 736 608 672 544 672 544 416ZM544 288L480 288 480 352 544 352 544 288Z"
      fill-opacity=".9"
      fill-rule="evenodd"
    />
  </svg>
  <!---->
  <span
    class="t-message-close"
  >
    <!---->
  </span>
</div>
`;

exports[`Message <slot> <icon> 1`] = `
<div
  class="t-message t-is-info"
>
  
  <svg
    class="t-icon t-icon-more"
    version="1.1"
    viewBox="0 0 1024 1024"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M576 192.00000000000003C576 227.34622400000003 547.346224 256.00000000000006 512 256.00000000000006 476.653776 256.00000000000006 448 227.34622400000003 448 192.00000000000003 448 156.65377600000002 476.653776 128.00000000000003 512 128.00000000000003 547.346224 128.00000000000003 576 156.65377600000002 576 192.00000000000003ZM576 512C576 547.346224 547.3462240000001 576 512 576 476.653776 576 448 547.346224 448 512 448 476.65377600000005 476.653776 448.00000000000006 512 448.00000000000006 547.346224 448.00000000000006 576 476.653776 576 512ZM576 832C576 867.3462208 547.3462240000001 896 512 896 476.653776 896 448 867.3462208 448 832 448 796.6537792 476.653776 768 512 768 547.3462240000001 768 576 796.6537792 576 832Z"
      fill-rule="evenodd"
    />
  </svg>
  
  <!---->
  <span
    class="t-message-close"
  >
    <!---->
  </span>
</div>
`;
